### ISSUE TITLE
[ERROR] CMAKE not find PHP_CONFIG_EXECUTABLE

### ISSUE THREAD

**Issue by iamrafaelmelo** [nano-interactive/cassandra-php-driver#79]
I am having problem when trying to compile the driver, when executing the command:

```sh
cmake --preset Release && cd out/Release && sudo ninja install
```

I've tried to install everything again, and it still didn't work, I was looking for solutions, I didn't find anything that could help me. The following error is returned:

```text
Preset CMake variables:

  CASS_CPP_STANDARD="17"
  CASS_USE_STATIC_LIBS="ON"
  CMAKE_BUILD_TYPE="Release"
  CMAKE_INSTALL_PREFIX="/home/<username>/Extensions/scylladb/out/install/Release"
  PHP_SCYLLADB_LIBSCYLLADB_STATIC="ON"
  PHP_SCYLLADB_LIBUV_STATIC="ON"

CMake Error at cmake/FindPHPConfig.cmake:34 (find_program):
  Could not find PHP_CONFIG_EXECUTABLE using the following names: php-config
Call Stack (most recent call first):
  CMakeLists.txt:37 (find_package
```

**On Docker**
<img width="563" alt="image" src="https://github.com/he4rt/scylladb-php-driver/assets/52138516/7fe5b872-3e73-4fbb-82e1-3f54e660eeec">

**PHP**: 8.1.2
**Return from the command `sudo update-alternatives --config php`**:
```text
There is only one alternative in link group php (providing /usr/bin/php): /usr/bin/php8.1
Nothing to configure.
```

**Return from the command `sudo update-alternatives --config php-config`**:
```text
There are 2 choices for the alternative php-config (providing /usr/bin/php-config).

  Selection    Path                         Priority   Status
------------------------------------------------------------
* 0            /usr/bin/php-config.default   100       auto mode
  1            /usr/bin/php-config.default   100       manual mode
  2            /usr/bin/php-config8.1        81        manual mode
```

**Comment by jkleinEV** [comment]
Hi

I encountered the same issue when compiling the driver.
To be able to continue, I modify the file `cmake/FindPHPConfig.cmake` as follow with the right path to my `php-config` file :

 ```
    #find_program(
    #        PHP_CONFIG_EXECUTABLE php-config
    #        HINTS ${hint}
    #        PATH_SUFFIXES bin
    #        NO_DEFAULT_PATH
    #        REQUIRED
    #)

    set(PHP_CONFIG_EXECUTABLE "/usr/bin/php-config")
```

It's not the good final solution but it allow me to continue the different step to compile the driver.

**Comment by iamrafaelmelo** [comment]
> Hi
> 
> I encountered the same issue when compiling the driver. To be able to continue, I modify the file `cmake/FindPHPConfig.cmake` as follow with the right path to my `php-config` file :
> 
> ```
>    #find_program(
>    #        PHP_CONFIG_EXECUTABLE php-config
>    #        HINTS ${hint}
>    #        PATH_SUFFIXES bin
>    #        NO_DEFAULT_PATH
>    #        REQUIRED
>    #)
> 
>    set(PHP_CONFIG_EXECUTABLE "/usr/bin/php-config")
> ```
> 
> It's not the good final solution but it allow me to continue the different step to compile the driver.

Thanks, I'll try that.

**Comment by iamrafaelmelo** [comment]
@jkleinEV after making this change, for some reason it is not finding the php development lib, even though it is installed:

```sh
Preset CMake variables:

  CASS_CPP_STANDARD="17"
  CASS_USE_STATIC_LIBS="ON"
  CMAKE_BUILD_TYPE="Release"
  CMAKE_INSTALL_PREFIX="/home/<username>/Extensions/scylladb/out/install/Release"
  PHP_SCYLLADB_LIBSCYLLADB_STATIC="ON"
  PHP_SCYLLADB_LIBUV_STATIC="ON"

-- Found php-config: /usr/bin/php-config
-- LIB: PHP_LIB-NOTFOUND # I put this message for debug
```

After moving `PHP_FOUND` out of the `PHP_LIB` condition, I can get through this:

**Before**:

```cmake
if (PHP_LIB)
    get_filename_component(PHP_LIB_DIR ${PHP_LIB} DIRECTORY)
    string(REGEX REPLACE "-I" "" PHP_INCLUDES "${PHP_INCLUDES}")
    string(REGEX REPLACE " " ";" PHP_INCLUDES "${PHP_INCLUDES}")
    set(PHP_LDFLAGS "-L${PHP_LIB_DIR} ${PHP_LIBS}")
    set(PHP_FOUND ON)
endif ()
```

**After**:

```cmake
if (PHP_LIB)
    get_filename_component(PHP_LIB_DIR ${PHP_LIB} DIRECTORY)
    string(REGEX REPLACE "-I" "" PHP_INCLUDES "${PHP_INCLUDES}")
    string(REGEX REPLACE " " ";" PHP_INCLUDES "${PHP_INCLUDES}")
    set(PHP_LDFLAGS "-L${PHP_LIB_DIR} ${PHP_LIBS}")
endif ()
set(PHP_FOUND ON)
```

But a new error is returned:

```sh
-- Build files have been written to: /home/<username>/Extensions/scylladb/out/Release
[1/332] Building CXX object CMakeFiles/ext_scylladb.dir/php_driver.cpp.o
FAILED: CMakeFiles/ext_scylladb.dir/php_driver.cpp.o 
/usr/bin/c++ -DCOMPILE_DL_CASSANDRA -DHAVE_CONFIG_H -DRELEASE -DSCYLLADB_STATIC -DUV_STATIC -Dext_scylladb_EXPORTS -I"/home/<username>/Extensions/scylladb/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I/home/<username>/Extensions/scylladb/include -I/home/<username>/Extensions/scylladb/out/Release -I/home/<username>/Extensions/scylladb/out/Release/_deps/libscylladb-src/include -I/home/<username>/Extensions/scylladb -I"/home/<username>/Extensions/scylladb/src/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/util/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/Type/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/DateTime/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/Database/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/Numbers/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/SSLOptions/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/Exception/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/Cluster/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/RetryPolicy/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I"/home/<username>/Extensions/scylladb/src/TimestampGenerator/-I/usr/include/php/20210902 -I/usr/include/php/20210902/main -I/usr/include/php/20210902/TSRM -I/usr/include/php/20210902/Zend -I/usr/include/php/20210902/ext -I/usr/include/php/20210902/ext/date/lib" -I/home/<username>/Extensions/scylladb/out/Release/_deps/libuv-src/include -O3 -DNDEBUG -std=gnu++20 -fPIC -fPIC -Wall -Wextra -Wno-long-long -Wno-deprecated-declarations -Wno-unused-parameter -Wno-unused-result -Wno-variadic-macros -Wno-extra-semi -pthread -O3 -MD -MT CMakeFiles/ext_scylladb.dir/php_driver.cpp.o -MF CMakeFiles/ext_scylladb.dir/php_driver.cpp.o.d -o CMakeFiles/ext_scylladb.dir/php_driver.cpp.o -c /home/<username>/Extensions/scylladb/php_driver.cpp
In file included from /home/<username>/Extensions/scylladb/src/DateTime/DateTimeInternal.h:3,
                 from /home/<username>/Extensions/scylladb/php_driver.cpp:18:
/home/<username>/Extensions/scylladb/ZendCPP/ZendCPP.hpp:3:10: fatal error: zend_API.h: No such file or directory
    3 | #include <zend_API.h>
      |          ^~~~~~~~~~~~
compilation terminated.
[10/332] Building C object _deps/libuv-build/CMakeFiles/uv.dir/src/uv-common.c.o
ninja: build stopped: subcommand failed.
```

A few weeks ago changes were made regarding this, I don't know what @malusev998 can help us with

**Comment by jkleinEV** [comment]
`php-config` is in /usr/bin/ directory on your environment ? If not, you have to adapt the "solution" that I gave.
If yes, I don't help you on your error if you say that the lib is already installed.

**Comment by iamrafaelmelo** [comment]
@jkleinEV I built the driver, however, that's all I got, going back 14 commits (even though my machine was all set up).

But thanks for the help anyway.

**Comment by iamrafaelmelo** [comment]
@malusev998 I'm going to keep this issue open because there have been changes to the CMAKE settings. I don't know if you can still compile the driver for the release version, if so, it's good to update the documentation.

**Comment by CodeLieutenant** [comment]
Yes, Currently there is a bug when finding a `php-config`, best solution for current building until it's fixed is to change `CMakePresets.json` like this

```
{
      "name": "Release",
      "displayName": "Release",
      "description": "",
      "generator": "Ninja",
      "environment": {},
      "binaryDir": "${sourceDir}/out/${presetName}",
      "cacheVariables": {
        "CUSTOM_PHP_CONFIG": "/path/to/php-config",
        ...restd
      }
    }
```

This is done for now in CI build for Github Actions. There is maybe another way, but I'm not sure this will work

```
cmake --preset Release -DCUSTOM_PHP_CONFIG="/path/to/php-config" && cd out/Release && sudo ninja install
```

I'll try to fix it next week

**Comment by CodeLieutenant** [comment]
Should be fixed in [54a2c93](https://github.com/he4rt/scylladb-php-driver/commit/54a2c936c88be954a1add4c6fd9ce1457d8b5ec4)

I'll be closing it now, free feel to reopen it if you still encounter the same error.
